src/c++/perf_analyzer/test_custom_load_manager.cc

```diff
@@ -0,0 +1,117 @@
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+#include "client_backend/client_backend.h"
+#include "constants.h"
+#include "custom_load_manager.h"
+#include "doctest.h"
+#include "request_rate_manager.h"
+
+namespace triton { namespace perfanalyzer {
+
+class TestCustomLoadManager : public CustomLoadManager {
+ public:
+  std::unique_ptr<std::chrono::nanoseconds>& gen_duration_{
+      RequestRateManager::gen_duration_};
+  std::vector<std::chrono::nanoseconds>& schedule_{
+      RequestRateManager::schedule_};
+  std::string& request_intervals_file_{
+      CustomLoadManager::request_intervals_file_};
+  std::vector<std::chrono::nanoseconds>& custom_intervals_{
+      CustomLoadManager::custom_intervals_};
+
+  cb::Error ReadTimeIntervalsFile(
+      const std::string& path,
+      std::vector<std::chrono::nanoseconds>* contents) override
+  {
+    return cb::Error::Success;
+  }
+};
+
+TEST_CASE("testing the InitCustomIntervals function")
+{
+  TestCustomLoadManager tclm{};
+
+  SUBCASE("no file provided")
+  {
+    cb::Error result{tclm.InitCustomIntervals()};
+
+    CHECK(result.Err() == SUCCESS);
+    CHECK(tclm.schedule_.size() == 1);
+    CHECK(tclm.schedule_[0] == std::chrono::nanoseconds(0));
+  }
+
+  SUBCASE("file provided")
+  {
+    tclm.request_intervals_file_ = "nonexistent_file.txt";
+    tclm.gen_duration_ = std::make_unique<std::chrono::nanoseconds>(350000000);
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(100000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(110000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(130000000));
+
+    cb::Error result{tclm.InitCustomIntervals()};
+
+    CHECK(result.Err() == SUCCESS);
+    CHECK(tclm.schedule_.size() == 5);
+    CHECK(tclm.schedule_[0] == std::chrono::nanoseconds(0));
+    CHECK(tclm.schedule_[1] == std::chrono::nanoseconds(100000000));
+    CHECK(tclm.schedule_[2] == std::chrono::nanoseconds(210000000));
+    CHECK(tclm.schedule_[3] == std::chrono::nanoseconds(340000000));
+    CHECK(tclm.schedule_[4] == std::chrono::nanoseconds(440000000));
```

Comment:
`custom_intervals_` only has 3 elements but because the `gen_duration_` is high enough, the code loops back through the beginning of `custom_intervals_` to keep filling the schedule. This test case shows that.

Reply:
> @tgerdesnv: If the schedule is A, B, C, then based on the gen duration we might do A, B, C, A, (0?), A, B, C, A.

Do you mean if the custom intervals is A, B, C, and then based on the gen duration we might do ...? Regardless, there would not be any extra 0s in the schedule. It starts at 0 and correctly loops through each custom interval, adding it to the previous schedule timestamp, until gen duration maximum is reached.

Reply:
I agree that it is going from custom_intervals to schedule properly, but it is possible to loop over the schedule multiple times. Gen_duration is currently only [2x](https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/perf_analyzer/request_rate_manager.cc#L86) the size of a single window, and generally we run for at least 3 windows, right?

If the user requests A, B, C, and that leads to a schedule of A, B, C, A,  then if we loop over the schedule multiple times we would end up A, B, C, A, A, B, C, A, which is probably not what the user wants. 

If that is in fact a bug, the fix is likely to just increase gen duration to be something like `window_size * max_windows`.

Reply:
I filed a ticket to follow up